File: src/main/java/net/pacalis/application/PodsWatcher.java
package net.pacalis.application;

import io.fabric8.kubernetes.api.model.Pod;
import io.fabric8.kubernetes.client.KubernetesClient;
import io.fabric8.kubernetes.client.KubernetesClientException;
import io.fabric8.kubernetes.client.Watch;
import io.fabric8.kubernetes.client.Watcher;
import io.quarkus.runtime.ShutdownEvent;
import io.quarkus.runtime.Startup;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.enterprise.context.ApplicationScoped;
import javax.enterprise.event.Observes;
import javax.inject.Inject;

@Startup
@ApplicationScoped
public class PodsWatcher {

    private static final Logger LOGGER = LoggerFactory.getLogger(PodsWatcher.class);

    private final Watch watch;

    public static class PodResourceWatcher implements Watcher<Pod> {
        @Override
        public void eventReceived(Action action, Pod pod) {
            // do something
        }

        @Override
        public void onClose(KubernetesClientException e) {
            // do something
        }
    }

    @Inject
    public PodsWatcher(KubernetesClient client) {
        this.watch = client.pods().watch(new PodResourceWatcher());
    }

    void onStop(@Observes ShutdownEvent ev) {
        LOGGER.info("The application is stopping...");
        watch.close();
    }
}

File: src/main/kotlin/net/pacalis/application/PodsWatcher.kt
package net.pacalis.application

import io.fabric8.kubernetes.api.model.Pod
import io.fabric8.kubernetes.client.KubernetesClient
import io.fabric8.kubernetes.client.KubernetesClientException
import io.fabric8.kubernetes.client.Watch
import io.fabric8.kubernetes.client.Watcher
import io.quarkus.runtime.ShutdownEvent
import io.quarkus.runtime.Startup
import org.slf4j.LoggerFactory
import javax.enterprise.context.ApplicationScoped
import javax.enterprise.event.Observes

@Startup
@ApplicationScoped
class PodsWatcher constructor(client: KubernetesClient, podHandler: PodHandler) {
    private val watch: Watch

    class PodResourceWatcher constructor(private val podHandler: PodHandler): Watcher<Pod?> {
        override fun eventReceived(action: Watcher.Action, pod: Pod?) {
            pod?.let {
                when(action) {
                    Watcher.Action.ADDED, Watcher.Action.MODIFIED -> podHandler.handle(pod)
                    else -> {}
                }
            }
        }

        override fun onClose(e: KubernetesClientException) {
            // Nothing to do
        }
    }

    fun onStop(@Observes ev: ShutdownEvent?) {
        LOGGER.info("Stop Pods watch ...")
        watch.close()
    }

    companion object {
        private val LOGGER = LoggerFactory.getLogger(PodsWatcher::class.java)
    }

    init {
        LOGGER.info("Synchronize pods...")
        client.pods().list().items.forEach {
            podHandler.handle(it)
        }
        watch = client.pods().watch(PodResourceWatcher(podHandler))
        LOGGER.info("Pods watch started.")
    }
}
